File: services/agent/src/routes/system.js
const express = require("express");
const si = require("systeminformation");

const router = express.Router();

router.get("/", async (req, res) => {
    try {
        const [
            cpuLoad,
            memory,
            battery,
            filesystem,
            network
        ] = await Promise.all([
            si.currentLoad(),
            si.mem(),
            si.battery(),
            si.fsSize(),
            si.networkStats()
        ]);

        res.json({
            cpu: {
                usage: cpuLoad.currentLoad.toFixed(2)
            },

            memory: {
                total: memory.total,
                used: memory.used,
                usage: ((memory.used / memory.total) * 100).toFixed(2)
            },

            battery: {
                percent: battery.percent,
                charging: battery.isCharging
            },

            disk: filesystem.map(disk => ({
                fs: disk.fs,
                used: disk.used,
                size: disk.size,
                usage: disk.use
            })),

            network: network
        });

    } catch (error) {
        console.error(error);

        res.status(500).json({
            error: "Failed to fetch system information"
        });
    }
});

module.exports = router;

File: services/agent/app/routes/system.py
import psutil
from fastapi import APIRouter

router = APIRouter()


@router.get("/system")
def get_system():
    memory = psutil.virtual_memory()
    disk = psutil.disk_usage("/")

    battery = psutil.sensors_battery()

    return {
        "cpu": round(psutil.cpu_percent(interval=0.5), 2),

        "memory": {
            "total": memory.total,
            "used": memory.used,
            "percent": memory.percent
        },

        "disk": {
            "total": disk.total,
            "used": disk.used,
            "percent": disk.percent
        },

        "battery": {
            "percent": battery.percent if battery else None,
            "plugged": battery.power_plugged if battery else None
        }
    }